src/EntityComponentManager_TEST.cc

```diff
@@ -1418,6 +1197,55 @@ TEST_P(EntityComponentManagerFixture, EachGetsNewOldRemove)
   EXPECT_EQ(0, removedCount<IntComponent>(manager));
 }
 
+//////////////////////////////////////////////////
+TEST_P(EntityComponentManagerFixture, EachAddRemoveComponent)
+{
+  // test calling ecm.Each on entities that have components added/removed
+  // frequently. This is common with *Cmd components
+
+  Entity e1 = manager.CreateEntity();
+  EXPECT_EQ(1u, manager.EntityCount());
+  EXPECT_EQ(0, eachCount<IntComponent>(manager));
+  auto comp = manager.Component<IntComponent>(e1);
+  EXPECT_EQ(nullptr, comp);
+
+  // add a component
+  manager.CreateComponent<IntComponent>(e1, IntComponent(123));
+  EXPECT_EQ(1, eachCount<IntComponent>(manager));
+  comp = manager.Component<IntComponent>(e1);
+  ASSERT_NE(nullptr, comp);
+  EXPECT_EQ(123, comp->Data());
+
+  // remove a component
+  manager.RemoveComponent(e1, IntComponent::typeId);
+  EXPECT_EQ(nullptr, manager.Component<IntComponent>(e1));
+  EXPECT_EQ(0, eachCount<IntComponent>(manager));
+
+  // add the same type of component back in
+  manager.CreateComponent<IntComponent>(e1, IntComponent(456));
+  EXPECT_EQ(1, eachCount<IntComponent>(manager));
+  comp = manager.Component<IntComponent>(e1);
+  ASSERT_NE(nullptr, comp);
+  EXPECT_EQ(456, comp->Data());
+
+  // remove the component again
+  manager.RemoveComponent(e1, IntComponent::typeId);
+  EXPECT_EQ(nullptr, manager.Component<IntComponent>(e1));
+  EXPECT_EQ(0, eachCount<IntComponent>(manager));
+
+  // add and remove the component before calling ecm.Each. This is to make sure
+  // that the views remove any entities in the toAdd queue if required
+  // components for entities in the toAdd queue are removed before calling Each
+  manager.CreateComponent<IntComponent>(e1, IntComponent(789));
+  comp = manager.Component<IntComponent>(e1);
+  ASSERT_NE(nullptr, comp);
+  EXPECT_EQ(789, comp->Data());
+  manager.RemoveComponent(e1, IntComponent::typeId);
```

Comment:
This specific test case is not supposed to call `eachCount` before removing again, because I am testing that the corresponding view's [`toAddEntities` queue](https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/include/ignition/gazebo/detail/BaseView.hh#L188-L190) is being updated properly: https://github.com/ignitionrobotics/ign-gazebo/blob/f8204deb4372993519e856ad020e2e59a5e3a158/src/EntityComponentManager_TEST.cc#L1330-L1334

Since a lot of this code is probably still pretty new to you, here's what is happening in this specific test code block:
1. call `CreateComponent` on an entity with a component type of `IntComponent`, which makes this entity valid for the `IntComponent` view and adds it to the view's `toAddEntities` queue (notice the `view->MarkEntityToAdd` call): https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/src/EntityComponentManager.cc#L603-L611
2. remove the component from the entity, which makes this entity no longer valid for the view, so the entity should be removed from the view's `toAddEntities` queue (notice the `NotifyComponentRemoval` call): https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/src/EntityComponentManager.cc#L381-L388
3. call `eachCount`, which internally calls `ecm.Each`. When `ecm.Each` is called, the view's `toAddEntities` queue should be empty, so the view shouldn't have any entities in it because it was initially empty and has no entities to be added to it: https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/include/ignition/gazebo/detail/EntityComponentManager.hh#L483-L492

So, as you can see, I am intentionally calling `eachCount` after `RemoveComponent` here because I want to make sure that the view empties its `toAddEntities` queue when needed. If I called `eachCount` before `RemoveComponent`, the view would clear the `toAddEntities` queue after adding all entities from the queue as a part of the `Each` call, and I wouldn't have a way to test this.

Does that make sense?